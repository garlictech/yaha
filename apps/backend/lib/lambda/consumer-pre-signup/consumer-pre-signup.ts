import * as fp from 'lodash/fp';
import { PreSignUpTriggerEvent, PreSignUpTriggerHandler } from 'aws-lambda';
import { flow, pipe } from 'fp-ts/lib/function';
import { CognitoIdentityServiceProvider } from 'aws-sdk';
import { combineLatest, defer, from, of, throwError } from 'rxjs';
import { catchError, map, switchMap, tap } from 'rxjs/operators';
import { throwIfEmptyValue } from '@bgap/shared/utils';

export interface Deps {
  cognitoidentityserviceprovider: CognitoIdentityServiceProvider;
}

export const handlerFn = (event: PreSignUpTriggerEvent) => (deps: Deps) => {
  console.debug('EVENT:', event);

  if (
    !(
      event.request.userAttributes.email ||
      event.request.userAttributes.phone_number
    )
  ) {
    throw new Error(
      'Bad request, either email or phone number must be provided',
    );
  }

  return pipe(
    [
      {
        AttributesToGet: ['email'],
        Filter: `email = "${event.request.userAttributes.email}"`,
      },
      {
        AttributesToGet: ['phone_number'],
        Filter: `phone_number = "${event.request.userAttributes.phone_number}"`,
      },
    ],
    params =>
      params.map(param => ({
        ...param,
        Limit: 2,
        UserPoolId: event.userPoolId,
      })),
    params =>
      params.map(param =>
        defer(() =>
          from(deps.cognitoidentityserviceprovider.listUsers(param).promise()),
        ).pipe(throwIfEmptyValue('BadDataFromCognito')),
      ),
    x => combineLatest(x),
    tap(x => console.debug('***1', x)),
    map(flow(result => result.map(res => res?.Users || []), fp.flatten)),
    tap(x => console.debug('***2', x)),
    switchMap(result =>
      result?.length ? throwError('UserAlreadyExists') : of(event),
    ),
    tap(x => console.debug('***2', x)),
    catchError(err => {
      console.error('ERROR:', JSON.stringify(err, null, 2));
      return throwError(err);
    }),
    tap(x => console.debug('***4', x)),
  ).toPromise();
};

const cognitoidentityserviceprovider = new CognitoIdentityServiceProvider({
  apiVersion: '2016-04-18',
  region: process.env.AWS_REGION,
});

export const handler: PreSignUpTriggerHandler = async (
  event,
  _context,
  _callback,
) =>
  await handlerFn(event)({
    cognitoidentityserviceprovider,
  });
