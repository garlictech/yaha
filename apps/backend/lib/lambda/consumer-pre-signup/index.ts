export { handler } from './consumer-pre-signup';
