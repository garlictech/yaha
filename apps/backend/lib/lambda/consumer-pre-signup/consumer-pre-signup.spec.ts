import { defer, from } from 'rxjs';
import { Deps, handlerFn } from './consumer-pre-signup';

// We use any-s to mock the system partially

const goodDeps: Deps = {
  cognitoidentityserviceprovider: {
    listUsers: jest
      .fn()
      .mockReturnValue({ promise: () => Promise.resolve({ Users: [] }) }),

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any,
};

const createGoodHandlerCall = (deps: Deps) =>
  defer(() =>
    from(
      handlerFn({
        request: {
          userAttributes: {
            email: 'a@a.hu',
            phone_numner: '+123456789',
          },
        },
        userPoolId: 'USERPOOL_ID',
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any)(deps),
    ),
  );

test('Handle the good case', done => {
  createGoodHandlerCall(goodDeps).subscribe(res => {
    expect(
      (goodDeps.cognitoidentityserviceprovider.listUsers as jest.Mock).mock
        .calls[0],
    ).toMatchSnapshot('parameters');

    expect(res).toMatchSnapshot('result');
    done();
  });
});

const listUserCaseDeps = (result: Promise<unknown>): Deps => ({
  ...goodDeps,
  cognitoidentityserviceprovider: {
    ...goodDeps.cognitoidentityserviceprovider,
    listUsers: jest.fn().mockReturnValue({
      promise: () => result,
    }),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any,
});

test('Handle listUser unknown error', done => {
  const deps = listUserCaseDeps(Promise.reject('UNKNOWN LISTUSERS ERROR'));

  createGoodHandlerCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});

test('Handle listUser returns bad data', done => {
  const deps = listUserCaseDeps(Promise.resolve(undefined));

  createGoodHandlerCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});

test('Handle multiple users found error', done => {
  const deps = listUserCaseDeps(Promise.resolve({ Users: [{}, {}] }));

  createGoodHandlerCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});

test('Handle user already exists error', done => {
  const deps = listUserCaseDeps(Promise.resolve({ Users: [{}] }));

  createGoodHandlerCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});
