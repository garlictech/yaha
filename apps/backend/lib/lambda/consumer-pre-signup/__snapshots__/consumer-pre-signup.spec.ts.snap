// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle listUser returns bad data 1`] = `[Error: BadDataFromCognito]`;

exports[`Handle listUser unknown error 1`] = `"UNKNOWN LISTUSERS ERROR"`;

exports[`Handle multiple users found error 1`] = `"UserAlreadyExists"`;

exports[`Handle the good case: parameters 1`] = `
Array [
  Object {
    "AttributesToGet": Array [
      "email",
    ],
    "Filter": "email = \\"a@a.hu\\"",
    "Limit": 2,
    "UserPoolId": "USERPOOL_ID",
  },
]
`;

exports[`Handle the good case: result 1`] = `
Object {
  "request": Object {
    "userAttributes": Object {
      "email": "a@a.hu",
      "phone_numner": "+123456789",
    },
  },
  "userPoolId": "USERPOOL_ID",
}
`;

exports[`Handle user already exists error 1`] = `"UserAlreadyExists"`;
